### DISCUSSION TITLE
using-lambda-at-edge-to-handle-angular-client-side-routing-with-s3-and-cloudfront/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#132] [Announcements]
# using-lambda-at-edge-to-handle-angular-client-side-routing-with-s3-and-cloudfront/

In this post I show how to use Lambda@Edge with a statically hosted site to intercept invalid requests caused by client-side routing and return index.html

https://andrewlock.net/using-lambda-at-edge-to-handle-angular-client-side-routing-with-s3-and-cloudfront/

**Comment by gao-artur** [comment]
Hi Andrew. Thank you for the lambda function, it is very helpful! Before we found this post, we used CustomErrorPage, similar to what you described in your first solution. And it created a bunch of problems! You better mark it as dangerous. Our distribution has two behaviors: `/` for the S3 resources and `/api` for the API. The 403/404 was overridden not only for the S3 but also for the API. So the API endpoint that is supposed to return a JSON document when the status code is 200 suddenly returned the HTML and broke our application. Not only that, this HTML was **cached** by the browser, hence replacing CustomErrorPage with the Lambda function wasn't enough. We had to instruct our clients to clear the cache before using the app again.

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at April 10 2019, 08:05, in reply to [DKarim](https://disqus.com/by/dkarim/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/132#discussioncomment-6056362)</em>

---

<p>Thanks! Interesting, I haven't seen that. Try adding "Transfer-Encoding" to the list of allowedHeaders  in `wrapAndFilterHeaders` and see if that solves it :)</p>

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at April 12 2019, 06:04, in reply to [DKarim](https://disqus.com/by/dkarim/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/132#discussioncomment-6056366)</em>

---

<p>Glad you got it sorted, and thanks for letting me know!</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at April 26 2020, 10:43, in reply to [Danny](https://disqus.com/by/danliebling/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/132#discussioncomment-6056372)</em>

---

<p>This isn't an issue I've seen, but I think you should be able to add 'transfer-encoding' to the list of allowed headers. If that works, I'll update the post :)</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at September 14 2019, 02:03, in reply to [David Mold](https://disqus.com/by/davidmold/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/132#discussioncomment-6056377)</em>

---

<p>I agree with you in principal, but the practicalities of that approach are tricky. Every deploy of the site may use different files (e.g for cashbusting etc, our angular build does this - the files have a suffix that changes each time). Plus you may need to maintain access to previous deploys (i.e. previous files)</p><p>So that means every deploy of the SPA site would have to deploy and publish a new Lambda too (and update the cloudfront distribution). The list of files in the lambda that are "ok" would need to be updated continually, and potentially grows (up to a point obviously).</p><p>if on the other hand, you have that a fixed set of file names, then yes, your approach may well be the easiest option :)</p>

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at September 15 2019, 08:26, in reply to [David Mold](https://disqus.com/by/davidmold/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/132#discussioncomment-6056379)</em>

---

<p>True, that would probably work in most cases :) Though not in my case, as routes contains dots. Another option would be to check the file extension. But again, has the potential to cause issues. The way I've presented is possibly "safest", if not the most performant.</p>

[a comment by an automated account is not shown]
